cookbooks/bcpc-hadoop/libraries/utils.rb

```diff
@@ -203,59 +203,84 @@ def set_hosts
     'hbase_head' => 'role[BCPC-Hadoop-Head-HBase]'
   }
 
+  service_recipe = {
+    'oozie_server' => 'recipe[bcpc-hadoop::oozie]'
+  }
+
   if node.run_state['cluster_def'].nil? then
     node.run_state['cluster_def'] = BACH::ClusterDef.new(node_obj: node)
   end
-  cd = node.run_state['cluster_def']
+  nodes = node.run_state['cluster_def'].fetch_cluster_def
+
+  # host search helper lambdas
+  runs_role = Proc.new { |host, role| host[:runlist].include? service_role[role] }
+  runs_recipe = Proc.new { |host, recipe| host[:runlist].include? service_recipe[recipe] }
+
+  # mapped host objects
+  to_host = Proc.new do |host| {
+    'hostname' => host[:fqdn],
+    'node_number' => host[:node_id],
+    'zookeeper_myid' => nil
+  } end
   
   node.default[:bcpc][:hadoop][:zookeeper][:servers] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['zookeeper'] }.map {
-      |hst| { 'hostname' => hst[:fqdn], 'node_number' => hst[:node_id], 'zookeeper_myid' => nil } }
+    nodes.select { |n| runs_role.call(n, 'zookeeper') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:jn_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['journal_node'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'journal_node') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:rm_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['resource_manager'] }.map {
-      |hst| { 'hostname' => hst[:fqdn], 'node_number' => hst[:node_id], 'zookeeper_myid' => nil } }
+    nodes.select { |n| runs_role.call(n, 'resource_manager') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:hs_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['job_history_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'job_history_server') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:dn_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hadoop_worker'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hadoop_worker') }.map { |n| to_host.call(n) }
 
   node.default[:bcpc][:hadoop][:hb_hosts] = 
-    cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hbase_head'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hbase_head') }.map { |n| to_host.call(n) }
 
   # different flavors of hive
   node.default[:bcpc][:hadoop][:hive_hosts] = 
-      cd.fetch_cluster_def.select { |hst| hst[:runlist].include? srvc2role['hive_server'] }.map {
-      |hst| { 'hostname' => hst[:fqdn]} }
+    nodes.select { |n| runs_role.call(n, 'hive_server') }.map { |n| to_host.call(n) }
 
-  # BCPC-Hadoop-Head-MapReduce
+  # (the BCPC-Hadoop-Head-MapReduce role nests the oozie recipe)
```

Comment:
I am not groking what the relevance of the MapReduce role is to anything here?

Reply:
The MapReduce role enforces that every node running a JobHistoryServer also runs an Oozie server. 
https://github.com/bloomberg/chef-bach/blob/master/stub-environment/roles/BCPC-Hadoop-Head-MapReduce.json

We might be better off changing this.